File: 102_binary-tree-level-order-traversal_BFS.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> order;
        if(root == nullptr) return order; 
        
        queue<pair<TreeNode*, int>> bfs;
        vector<int> same_lvl;    
        bfs.push(pair<TreeNode*, int>(root, 1));
        int current_lvl = 1;
        while(!bfs.empty()){
            pair<TreeNode*, int> front = bfs.front();
            bfs.pop();
            // Check level
            if(front.second == current_lvl)
                // the front node is still at the same level
                same_lvl.push_back(front.first->val);
            else{
                // the front node is at a new level
                // push the previous level into answer
                order.push_back(same_lvl);
                same_lvl.clear();
                same_lvl.push_back(front.first->val);
                current_lvl++;
            }
            // Push left child into the queue
            if(front.first->left != nullptr)
                bfs.push(pair<TreeNode*, int>(front.first->left, front.second+1));
            // Push right child into the queue
            if(front.first->right != nullptr)
                bfs.push(pair<TreeNode*, int>(front.first->right, front.second+1));
        }
        // Push the vector of last level into answer
        order.push_back(same_lvl);
        return order;
    }
};

File: 102_binary-tree-level-order-traversal_bfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None: return []
        queue, order = collections.deque(), []
        queue.append(root)
        while queue:
            temp = []
            size = len(queue)
            for i in range(size):
                top = queue.popleft()
                temp.append(top.val)
                if top.left is not None: queue.append(top.left)
                if top.right is not None: queue.append(top.right)
            order.append(temp)
        return order
